File: src/functions/zend/connections.ts
import { Connection } from "./connection";
import { ZendRequest } from "./types";
import { _connect } from "../functions";

export class Connections {
  connections: Map<string, Connection>;

  constructor() {
    this.connections = new Map<string, Connection>();
  }

  add(handle: string, connection: Connection): void {
    this.connections.set(handle, connection);
  }

  async sendToAll(request: ZendRequest): Promise<void> {
    const sendOne = async ([, connection]: [string, Connection]) => {
      if (connection.getClient() === null) {
        connection.setClient(await _connect(connection.getHost(), connection.getLocalPort()));
      }
      await connection.getClient().sendRequest(request);
    };
    await Promise.all(Array.from(this.connections).map(sendOne));
  }
}

File: zend/zend-master/Java/src/zend/Connections.java
package zend;

import org.bson.Document;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import static zend.Sydx._connect;

public class Connections {
  private Map<String, Connection> connections;

  public Connections() {
    connections = new HashMap<>();
  }

  public synchronized void add(String handle, Connection connection) {
    connections.put(handle, connection);
  }

  public synchronized void sendToAll(Document request) throws IOException {
    for (Map.Entry<String, Connection> connectionEntry : connections.entrySet()) {
      Connection connection = connectionEntry.getValue();
      if (connection.getClient() == null){
        connection.setClient(_connect(connection.getHost(), connection.getLocalPort()));
      }
      connection.getClient().sendRequest(request);
    }
  }
}